algebra-core/src/main/scala/algebra/lattice/BoundedJoinSemilattice.scala

```diff
@@ -24,15 +24,12 @@ package lattice
 
 import scala.{specialized => sp}
 
-trait BoundedJoinSemilattice[@sp(Int, Long, Float, Double) A] extends Any with JoinSemilattice[A] { self =>
+trait BoundedJoinSemilattice[@sp(Int, Long, Float, Double) A] extends Any with JoinSemilattice[A] {
   def zero: A
   def isZero(a: A)(implicit ev: Eq[A]): Boolean = ev.eqv(a, zero)
 
   override def joinSemilattice: BoundedSemilattice[A] =
-    new BoundedSemilattice[A] {
-      def empty: A = self.zero
-      def combine(x: A, y: A): A = join(x, y)
-    }
+    BoundedSemilattice.instance(zero, join)
```

Comment:
This isn't really a SAM, so isn't it subject to my same concern from https://github.com/typelevel/cats/pull/3871#issuecomment-1116057617?

Reply:
I think the concern was when there are more than one lambda

Reply:
Hmm, not sure why this is different. I think it is storing two references (zero and join) where it previously stored one.

Reply:
Right, but is that really a problem when before we used to have an entire new class instead of an additional reference?

Reply:
Except, you have one class per JVM vs an extra reference per instance :) FTR I don't particularly care about this, I'm just unconvinced whether we are really optimizing anything here. See my comment in https://github.com/typelevel/cats/pull/3871#issuecomment-979604249.

> Additionally, isn't this de-optimizing the current implementations? By replacing dedicated classes with lambdas and thus adding a level of indirection and a larger per-instance footprint. Like, these are micro-optimizations, but isn't saving a few megabytes? :)

Reply:
Well if it's not worth it then we can also close this PR and the related issue.
One minor advantage is that the definitions get a bit more concise.